"no-undef" rule and "browser" environment acting strange with regard to identifier case

**What version are you using?**
```bash
$ eslint -v
v1.3.1
```

**What did you do?**
```bash
$ echo "var Document = {};" | eslint -f json --stdin --no-eslintrc --env "browser" --rule '{"no-undef": 2}'; echo $?
```
**What happened?**
```json
[{"filePath":"<text>","messages":[{"ruleId":"no-undef","severity":2,"message":"\"Document\" is read only.","line":1,"column":5,"nodeType":"Identifier","source":"var Document = {};"}],"errorCount":1,"warningCount":0}]
1
```

**What did you expect to happen?**
```json
[{"filePath":"<text>","messages":[],"errorCount":0,"warningCount":0}]
0
```

--

Oddly enough, if one adds the "commonjs" environment to the command invocation, the expected output is shown.  I have no idea why adding the "commonjs" environment makes this go away from the CLI, but my actual use-case is invoking ESLint with an hierarchical inherited `.eslintrc` configuration via webpack, and the file where this issue came up is configured to lint with both the "browser" and the "commonjs" environments, and I still receive the error there.

I think it's safe to ignore this CLI w/o `.eslintrc` vs webpack w/ `.eslintrc` oddity and just fix the case-sensitivity issue with "browser" environment/globals/the no-undef rule.

Actually `no-redeclare` and `no-shadow` are handling the extra scope as global.

```js
/* eslint-env commonjs */
/* eslint no-redeclare:[2,{builtinGlobals:true}], no-shadow:[2,{builtinGlobals:true}] */
var Object = 0;
```

```
test.js
  3:5  error  Object is already defined  no-redeclare

✖ 1 problem (1 error, 0 warnings)
```

But also I understand the request in #3390...
Built-in globals includes many general names, especially in browsers.
So if somebody are 100% sure extra scope is made, they expect to not handle it as redeclaration.

(Ah, this is "difficult arrangement" I said before)

--- a/lib/config.js
+++ b/lib/config.js
@@ -70,6 +70,44 @@ function isObject(item) {
 }
 
 /**
+ * Creates an environment config based on the specified environments.
+ * @param {Object<string,boolean>} envs The environment settings.
+ * @returns {Object} A configuration object with the appropriate rules and globals
+ *      set.
+ * @private
+ */
+function createEnvironmentConfig(envs) {
+
+    var envConfig = {
+        globals: {},
+        env: envs || {},
+        rules: {},
+        ecmaFeatures: {}
+    };
+
+    if (envs) {
+        Object.keys(envs).filter(function(name) {
+            return envs[name];
+        }).forEach(function(name) {
+            var environment = environments[name];
+
+            if (environment) {
+
+                if (environment.globals) {
+                    assign(envConfig.globals, environment.globals);
+                }
+
+                if (environment.ecmaFeatures) {
+                    assign(envConfig.ecmaFeatures, environment.ecmaFeatures);
+                }
+            }
+        });
+    }
+
+    return envConfig;
+}
+
+/**
  * Read the config from the config JSON file
  * @param {string} filePath the path to the JSON config file
  * @returns {Object} config object
@@ -175,6 +213,10 @@ function loadConfig(configToLoad) {
 
         }
 
+        if (config.env) {
+            // Merge in environment-specific globals and ecmaFeatures.
+            config = util.mergeConfigs(createEnvironmentConfig(config.env), config);
+        }
 
     }
 
@@ -291,44 +333,6 @@ function getLocalConfig(thisConfig, directory) {
     return found ? config : util.mergeConfigs(config, getPersonalConfig());
 }
 
-/**
- * Creates an environment config based on the specified environments.
- * @param {Object<string,boolean>} envs The environment settings.
- * @returns {Object} A configuration object with the appropriate rules and globals
- *      set.
- * @private
- */
-function createEnvironmentConfig(envs) {
-
-    var envConfig = {
-        globals: {},
-        env: envs || {},
-        rules: {},
-        ecmaFeatures: {}
-    };
-
-    if (envs) {
-        Object.keys(envs).filter(function(name) {
-            return envs[name];
-        }).forEach(function(name) {
-            var environment = environments[name];
-
-            if (environment) {
-
-                if (environment.globals) {
-                    assign(envConfig.globals, environment.globals);
-                }
-
-                if (environment.ecmaFeatures) {
-                    assign(envConfig.ecmaFeatures, environment.ecmaFeatures);
-                }
-            }
-        });
-    }
-
-    return envConfig;
-}
-
 //------------------------------------------------------------------------------
 // API
 //------------------------------------------------------------------------------
@@ -412,44 +416,37 @@ Config.prototype.getConfig = function(filePath) {
     // Step 2: Create a copy of the baseConfig
     config = util.mergeConfigs({parser: this.parser}, this.baseConfig);
 
-    // Step 3: Merge in environment-specific globals and rules from .eslintrc files
-    config = util.mergeConfigs(config, createEnvironmentConfig(userConfig.env));
-
-    // Step 4: Merge in the user-specified configuration from .eslintrc and package.json
+    // Step 3: Merge in the user-specified configuration from .eslintrc and package.json
     config = util.mergeConfigs(config, userConfig);
 
-    // Step 5: Merge in command line config file
+    // Step 4: Merge in command line config file
     if (this.useSpecificConfig) {
         debug("Merging command line config file");
-
-        if (this.useSpecificConfig.env) {
-            config = util.mergeConfigs(config, createEnvironmentConfig(this.useSpecificConfig.env));
-        }
 
         config = util.mergeConfigs(config, this.useSpecificConfig);
     }
 
-    // Step 6: Merge in command line environments
+    // Step 5: Merge in command line environments
     debug("Merging command line environment settings");
     config = util.mergeConfigs(config, createEnvironmentConfig(this.env));
 
-    // Step 7: Merge in command line rules
+    // Step 6: Merge in command line rules
     if (this.options.rules) {
         debug("Merging command line rules");
         config = util.mergeConfigs(config, { rules: this.options.rules });
     }
 
-    // Step 8: Merge in command line globals
+    // Step 7: Merge in command line globals
     config = util.mergeConfigs(config, { globals: this.globals });
 
-    // Step 9: Merge in command line plugins
+    // Step 8: Merge in command line plugins
     if (this.options.plugins) {
         debug("Merging command line plugins");
         pluginConfig = getPluginsConfig(this.options.plugins);
         config = util.mergeConfigs(config, { plugins: this.options.plugins });
     }
 
-    // Step 10: Merge in plugin specific rules in reverse
+    // Step 9: Merge in plugin specific rules in reverse
     if (config.plugins) {
         pluginConfig = getPluginsConfig(config.plugins);
         config = util.mergeConfigs(pluginConfig, config);

--- a/tests/fixtures/config-hierarchy/overwrite-ecmaFeatures/.eslintrc
+++ b/tests/fixtures/config-hierarchy/overwrite-ecmaFeatures/.eslintrc
@@ -1,0 +1,5 @@
+{
+    "ecmaFeatures": {
+        "globalReturn": false
+    }
+}

--- a/tests/fixtures/config-hierarchy/overwrite-ecmaFeatures/child/.eslintrc
+++ b/tests/fixtures/config-hierarchy/overwrite-ecmaFeatures/child/.eslintrc
@@ -1,0 +1,5 @@
+{
+    "env": {
+        "commonjs": true
+    }
+}

--- a/tests/lib/config.js
+++ b/tests/lib/config.js
@@ -35,6 +35,10 @@ function assertConfigsEqual(actual, expected) {
         assert.deepEqual(actual.env, expected.env);
     }
 
+    if (actual.ecmaFeatures && expected.ecmaFeatures) {
+        assert.deepEqual(actual.ecmaFeatures, expected.ecmaFeatures);
+    }
+
     if (actual.globals && expected.globals) {
         assert.deepEqual(actual.globals, expected.globals);
     }
@@ -1106,6 +1110,21 @@ describe("Config", function() {
                 var expected = configHelper.getConfig(file2);
 
                 assert(!("quotes" in expected.rules), "shared config should not be clobbered");
+            });
+        });
+
+        describe("with env in a child configuration file", function() {
+            it("should overwrite ecmaFeatures of the parent with env of the child", function() {
+                var config = new Config();
+                var targetPath = getFixturePath("overwrite-ecmaFeatures", "child", "foo.js");
+                var expected = {
+                    rules: {},
+                    env: {commonjs: true},
+                    ecmaFeatures: {globalReturn: true}
+                };
+                var actual = config.getConfig(targetPath);
+
+                assertConfigsEqual(actual, expected);
             });
         });
     });

